### PR TITLE
Vehicle racks keep vehicle zones on merge and split

### LINKED ISSUE #27645: Bike Racks remove Vehicle Zones
**Describe the bug**  
Vehicle Zones get removed whenever it's attached to a bike rack. For example, if I have the Unsorted tag on a shopping cart, and I attach said cart to a Bike Rack, the zone is lost and will need to be re-placed again to make it react to other zones.

**To Reproduce**  
1. Place a Vehicle Zone on your Shopping Cart
2. Attach it to your Vehicle's Bike Rack.
3. Watch as the Zone disappears from said Shopping Cart.

**Expected behavior**  
The Vehicle Zone should be persistent and static and not temporary. It should stay on the Shopping Cart. You should not have to re-add the Vehicle Zone to said cart every time you load or take off a vehicle from a Bike Rack

**Versions and configuration(please complete the following information):**
 - OS: Windows 10 64-bit
 - Game Version 0.C-36419-gd7a0f0c (tiles)
 - Graphics version Tiles
 - Mods loaded DDA

### PR DESCRIPTION
**Pull request by MetroidHunter**
#### Summary
SUMMARY: Bugfixes "Retaining vehicle zones when merging and splitting vehicles"

#### Purpose of change
Fixes #27645 

#### Describe the solution
Looking at how vehicles merge, zones were not being considered. The vehicle to merge is broken into mount points with parts, those points are cached into a list of mappings and those mappings are then used to add the parts and points to the other vehicle. At the point where we add all the parts to the new mount point on the other vehicle, we now check for any zones on the vehicle to merge and then add them to the other vehicle in the same locations they were for the vehicle to merge but now on the new mount points. The vehicle to merge is then destroyed which removes the zones from that vehicle completing the transition.

Regarding splitting, the logic is setup to create a new vehicle based on the existing vehicle and relative mount points. The new vehicle is created, for the most part, all at once. There was some existing zone logic here but it was naive and did not take into account the zone caching and multiple structures that track the zones. We now, while adding individual parts to the new vehicle, fully use the zone manager to remove the existing zone from the existing vehicle. We then reuse a structure that was tracking the removed zones to add them to the new vehicle after we process all the parts. We wait until we process all the parts because we only need to attach the zones to a mount point and if we do that in the loop processing the zones, we will end up creating multiple zones on a single mount point, one for each part at that location.

I renamed `find_rackable_vehicles` to `try_to_rack_nearby_vehicle` in order to better describe what that was doing.

I did some light refactoring, moving variables around in the merge logic to correspond to where they are used for easier reading. I also added a bit of documentation to explain the logic a bit since its currently a tad confusing.

### CODE REVIEW

**Comment by kevingranade** [comment]
Bit of an error.  Loaded a motorcycle with a zone on the cargo box on the back onto a SUV with a bike rack, then drove around and unracked it.  I got several errors:
Tried to remove a zone from an unloaded vehicle
```
        Attempting to repeat stack trace using debug symbols...
        debug_write_backtrace(std::ostream&)
        .../src/debug.cpp:591
        DebugLog(DebugLevel, DebugClass)
        .../src/debug.cpp:750
        realDebugmsg(char const*, char const*, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
        .../src/debug.cpp:90
        std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()
        /usr/include/c++/7/bits/basic_string.h:647
        void realDebugmsg<>(char const*, char const*, char const*, char const*)
        .../src/debug.h:76
        zone_manager::remove(zone_data&)
        .../src/clzones.cpp:691
        vehicle::split_vehicles(std::vector<std::vector<int, std::allocator<int> >, std::allocator<std::vector<int, std::allocator<int> > > > const&, std::vector<vehicle*, std::allocator<vehicle*> > const&, std::vector<std::ve
ctor<point, std::allocator<point> >, std::allocator<std::vector<point, std::allocator<point> > > > const&)
        .../src/vehicle.cpp:1885
        vehicle::remove_carried_vehicle(std::vector<int, std::allocator<int> > const&)
        .../src/vehicle.cpp:1677
        vehicle::use_bike_rack(int)
        .../src/vehicle_use.cpp:1347
        interact_with_vehicle
        .../src/pickup.cpp:207
        Pickup::pick_up(tripoint const&, int)
        .../src/pickup.cpp:645
        game::examine(tripoint const&)
        .../src/game.cpp:6415
        cata::optional<tripoint>::~optional()
        .../src/optional.h:99
        game::examine()
        .../src/game.cpp:6326
        game::handle_action()
        .../src/handle_action.cpp:1449
        game::do_turn()
        .../src/game.cpp:1569
        main
        .../src/main.cpp:673
        __libc_start_main
        _start
```
And then upon examination, the bicycle cargo box has 4 copies of the zone, and the driver-side door of the SUV also had a copy.

Off the top of my head I'd guess the issue is it stored the zone after finding it, then re-applied it repeatedly while iterating.  Probably it needs to be cleared somewhere.

**Comment by MetroidHunter** [comment]
I bet its creating a zone per mount point of the racked vehicle. Investigating now

**Comment by kevingranade** [comment]
After a bunch of debugging, I think I have the cause, but I don't have a fix yet.

It looks like the absolute location of the loot zone on the carried vehicle is not getting updated as the vehicle moves, so when the split code tries to remove the zone after copying it to the newly split vehicle, it fails because it can't find the vehicle it's attached to.  As a result, each subsequent trip through the loop that is iterating over the parts of the current mount point inserts another copy of the zone into the new vehicle and complains about being unable to remove the old one.

**Comment by kevingranade** [comment]
You can have an arbitrary number of zones in a single square, the code doesn't currently copy all of them over, just the first.

**Comment by MetroidHunter** [comment]
After doing a bunch of reading about multimaps, which I have not previously used, I've got the multiple zones working and correctly splitting and merging the vehicles regardless of which vehicle has moved before or after being merged or split.

The zones interactions with vehicles seems to be lazy in that you mark the vehicle has having dirty zones and then next time an interaction with the zone_manager happens, the zones will be rebuilt (like using the zones or opening the menu). Because of that, we are free to just remove from loot_zones without having to call remove directly or having to eagerly refresh the zone positions.

File: src/vehicle.cpp [hunk 1]
```diff
@@ -1843,7 +1879,10 @@ bool vehicle::split_vehicles( const std::vector<std::vector <int>> &new_vehs,
             const auto lz_iter = loot_zones.find( cur_mount );
             if( lz_iter != loot_zones.end() ) {
                 new_zones.emplace( new_mount, lz_iter->second );
```
[the remark this answers is not in the document]
Comment (MetroidHunter) [reply src/vehicle.cpp]: It turns out we dont need to refresh the zones if we allow the updates to be lazy which seems to be the original intention in the vehicle logic. It was using the eager remove from the zone_manager that caused the refresh to be necessary there

**Comment by kevingranade** [comment]
Merged to 0.D and master.

**Pull request closed without merging**